### ISSUE TITLE
Expected X-Hasura-Base-Domain header

### ISSUE THREAD

**Issue by CYBONYMOUS** [hasura/docs#365]
I am using the internal endpoints to contact the auth service from within another service and the headers I sent are:
headers = {
        "Content-Type": "application/json",
        "X-Hasura-User-Id": '0',
        "X-Hasura-Role": "admin"
         }
as mentioned in the docs.
How ever I got an error saying:
b'{"code":"unexpected-state","message":"Expected X-Hasura-Base-Domain header","detail":null}'
I've searched for this error and was unable to find it.
Please look into this.

**Comment by dsandip** [comment]
@CYBONYMOUS : You will need to make the call to the external endpoint for i.e. auth.<cluster-name>.hasura-app.io/<endpoint> when making a call from a custom service or proxy all headers coming from the gateway (when your custom service is called), and proxy back all headers received from auth with the response. Can you try either of these methods and confirm if you are able to resolve this.

**Comment by dsandip** [comment]
@ecthiender: Once @CYBONYMOUS confirms, we can change this issue into for auth docs to go into more detail as we discussed.

**Comment by CYBONYMOUS** [comment]
Should I make a request to the external endpoint with my auth token and send the headers that I get in my response to the internal endpoints?

**Comment by dsandip** [comment]
@CYBONYMOUS : Almost right. Yes, you will need to make a call from your customer service to the external endpoint (with a token of course) and then send back the headers that you get from auth in the response from your service. @ecthiender: Confirm this?

**Comment by 0x777** [comment]
@CYBONYMOUS Hey, so the auth service requires this header to be present `X-Hasura-Base-Domain` (as it needs to deal with cookies). If you are trying to contact auth's internal end points, you should set this header to the value that gateway sets in the request to your microservice. If this is your 

```
Client ----> Gateway (adds X-hasura-base-domain and others) -----> Your microservice (add x-hasura-base-domain given by gateway and other headers) -----> auth's internal service
```

**Comment by CYBONYMOUS** [comment]
After making a request to the login endpoint, all the headers I receive from the response are
{
    "auth_token": "",
    "username": "johnsmith",
    "hasura_id": 2,
    "hasura_roles": [
        "user"
    ]
}
I don't find X-hasura-base-domain

**Comment by 0x777** [comment]
> I am using the internal endpoints to contact the auth service from within another service and the headers I sent are:
headers = {
"Content-Type": "application/json",
"X-Hasura-User-Id": '0',
"X-Hasura-Role": "admin"
}
as mentioned in the docs.
How ever i got a error saying:
b'{"code":"unexpected-state","message":"Expected X-Hasura-Base-Domain header","detail":null}'
I've searched for this error and was unable to find it.
Please look into this.

In your custom service, read the header `X-Hasura-Base-Domain` from the request (the gateway sets this) and then when making the request to auth's internal endpoints add this header

**Comment by CYBONYMOUS** [comment]
Hey, in the above message, I posted the response I received after making a request to the login endpoint and I find no header named `X-Hasura-Base-Domain` in the response headers

**Comment by ecthiender** [comment]
Hey @CYBONYMOUS , read the sentences above carefully. @0x777 is saying the request that *your* microservice receives (from the frontend) *will have* the headers (as it is added by the gateway).

And you are saying you are not receiving the headers from auth login endpoint. These are 2 different things.

When *your* code receives the request (it will have all the `x-hasura-*` headers), forward all of them to auth on the *login request*. Auth will return some response headers, send them back as it is to clients.

Basically, proxy all headers that you receive on request to auth. Proxy back all response headers your receive from auth back to the client.

**Comment by ecthiender** [comment]
@CYBONYMOUS or use the external endpoint. Make calls to `auth.cluster.hasura-app.io` directly. And don't use the internal endpoint at all.

**Comment by CYBONYMOUS** [comment]
But the auth token keeps changing

**Comment by ecthiender** [comment]
Closed in favour of #394
